pkg/config/config_test.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config_test
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/kinvolk/lokomotive/pkg/config"
+)
+
+//nolint:funlen,gocognit
+func TestReadHCL(t *testing.T) {
+	tests := []struct {
+		desc    string
+		conf    []byte
+		vars    []byte
+		wantErr bool
+	}{
+		{
+			desc: "Cluster with vars",
+			conf: []byte(`variable "cluster_name" {}
+
+cluster "aws" {
+  asset_dir        = pathexpand("./assets")
+  cluster_name     = var.cluster_name
+  controller_count = 1
+  dns_zone         = "example.com"
+  dns_zone_id      = "fake"
+  ssh_pubkeys      = ["stuff"]
+  
+  worker_pool "my-wp-name" {
+    count         = 3
+    instance_type = "fake"
+    ssh_pubkeys   = ["stuff"]
+  }
+}`),
+			vars: []byte(`cluster_name = "test-cluster"`),
+		},
+		{
+			desc: "Cluster without vars",
+			conf: []byte(`cluster "aws" {
+  asset_dir        = pathexpand("./assets")
+  cluster_name     = var.cluster_name
+  controller_count = 1
+  dns_zone         = "example.com"
+  dns_zone_id      = "fake"
+  ssh_pubkeys      = ["stuff"]
+  
+  worker_pool "my-wp-name" {
+    count         = 3
+    instance_type = "fake"
+    ssh_pubkeys   = ["stuff"]
+  }
+}`),
+		},
+		{
+			desc: "Empty config",
+			conf: []byte(``),
+			// TODO: Is it OK that an empty config doesn't generate errors?
+		},
+		{
+			desc: "Empty vars",
+			conf: []byte(`variable "cluster_name" {}
+cluster "aws" {}`),
+			vars: []byte(``),
+		},
+		{
+			desc:    "Malformed config",
+			conf:    []byte(`oops`),
+			wantErr: true,
+		},
+		{
+			desc:    "Non-existent block",
+			conf:    []byte(`foo "bar" {}`),
+			wantErr: true,
+		},
+	}
+
+	d, err := ioutil.TempDir("", "lokoctl-tests-")
+	if err != nil {
+		t.Fatalf("Could not create temporary directory: %v", err)
+	}
+
+	t.Cleanup(func() {
+		if err := os.RemoveAll(d); err != nil {
+			t.Logf("Could not remove temporary directory %q: %v", d, err)
+		}
+	})
+
+	for _, test := range tests {
+		// Pin test to avoid scope bugs: https://github.com/kyoh86/scopelint
```

Comment:
This is a standard thing to do, I think we can remove this comment.

Reply:
Is it? I've encountered it only recently.

Reply:
Well, we do it only in couple of places, so I guess it's fine to leave it :smile: 
```
$ git grep -A 1 range pkg/**/*.go cmd/**/*.go test/**/*.go | grep -v range | grep := | grep -v '(' | grep -v '{'
test/monitoring/components_alerts_test.go-              tc := tc
test/monitoring/components_metrics_test.go-             tc := tc
test/monitoring/monitoring_test.go-             test := test
```

Reply:
I would actually rather do without the comment but I thought people would be confused by this strange assignment.

Reply:
I removed these comments. The linter complains when you don't pin these vars so I think it's obvious enough.